transaction.go

```diff
@@ -153,49 +289,121 @@ func StartFunc(service, user string, handler func(Style, string) (string, error)
 //
 // All application calls to PAM begin with Start*. The returned
 // transaction provides an interface to the remainder of the API.
+//
+// It's responsibility of the Transaction owner to release all the resources
+// allocated underneath by PAM by calling End() once done.
+//
+// It's not advised to End the transaction using a runtime.SetFinalizer unless
+// you're absolutely sure that your stack is multi-thread friendly (normally it
+// is not!) and using a LockOSThread/UnlockOSThread pair.
 func StartConfDir(service, user string, handler ConversationHandler, confDir string) (*Transaction, error) {
 	if !CheckPamHasStartConfdir() {
-		return nil, errors.New("StartConfDir() was used, but the pam version on the system is not recent enough")
+		return nil, NewTransactionError(ErrSystem,
+			errors.New("StartConfDir() was used, but the pam version on the system is not recent enough"))
 	}
 
 	return start(service, user, handler, confDir)
 }
 
+type panicHandlerFunc func(any)
+
+var defaultPanicHandler = func(v any) { panic(v) }
+
+var panicHandler panicHandlerFunc = defaultPanicHandler
+
 func start(service, user string, handler ConversationHandler, confDir string) (*Transaction, error) {
 	switch handler.(type) {
 	case BinaryConversationHandler:
 		if !CheckPamHasBinaryProtocol() {
-			return nil, errors.New("BinaryConversationHandler() was used, but it is not supported by this platform")
+			return nil, NewTransactionError(ErrSystem,
+				errors.New("BinaryConversationHandler() was used, but it is not supported by this platform"))
 		}
 	}
 	t := &Transaction{
 		conv: &C.struct_pam_conv{},
 		c:    cgo.NewHandle(handler),
 	}
+
+	callers := make([]uintptr, 10)
+	haveCallerInfo := runtime.Callers(2, callers) > 0
+	runtime.SetFinalizer(t, func(t *Transaction) {
```

Comment:
Should we remove this and make calling `End` a hard requirement? After we land this patch I'll tag 2.0. I think a note in the readme about changes between 1.0 and 2.0 would suffice rather than warnings in the function documentation about calling `runtime.SetFinalizer`.

Reply:
Well, it's a safety check to ensure `End()` got called, panic'ing in case, I though was a good way to catch potential errors in tests.

But we can drop if you think it's something "over-engineered".

Reply:
I'm pretty open to your thinking on this, but my vote is to remove it. When I originally wrote the finalizer stuff I was pretty new to Go. I think using a finalizer here was a mistake.

Reply:
No, I mean, the way I'm using here is just to panic on people mistakes. Bsaically an `assert`, while it won't do anything with the module itself.

Reply:
Ok, I'll maybe add this in the tests to double-check things.

Reply:
Ack, done.